contractsV5/KyberTradeLogic.sol

```diff
@@ -0,0 +1,414 @@
+        if (src == dest || reserveArr.length == 0) return (IKyberReserve(0), PRECISION, false);
+
+        if (reserveArr.length == 0) return (IKyberReserve(0), 0, false);
+
+        uint[] memory rates = new uint[](reserveArr.length);
+        uint[] memory reserveCandidates = new uint[](reserveArr.length);
+        uint destAmount;
+        uint srcAmountWithFee;
+
+        for (uint i = 0; i < reserveArr.length; i++) {
+            reserve = reserveArr[i];
+            //list all reserves that support this token.
+            isPayingFees = isFeePayingReserve[address(reserve)];
+            //for ETH -> token paying reserve, takerFee is specified in amount
+            srcAmountWithFee = ((src == ETH_TOKEN_ADDRESS) && isPayingFees) ? srcAmount - takerFee : srcAmount;
+            rates[i] = reserve.getConversionRate(
+                src,
+                dest,
+                srcAmountWithFee,
+                block.number);
+
+            destAmount = srcAmountWithFee * rates[i] / PRECISION;
+            //for token -> ETH paying reserve, takerFee is specified in bps
+            destAmount = (dest == ETH_TOKEN_ADDRESS && isPayingFees) ? destAmount * (BPS - takerFee) / BPS : destAmount;
+
+            if (destAmount > bestReserve.destAmount) {
+                //best rate is highest rate
+                bestReserve.destAmount = destAmount;
+                bestReserve.index = i;
+            }
+        }
+
+        if(bestReserve.destAmount == 0) return (IKyberReserve(0), 0, false);
+
+        reserveCandidates[0] = bestReserve.index;
+
+        // if this reserve pays fee its actual rate is less. so smallestRelevantRate is smaller.
+        bestReserve.destAmount = bestReserve.destAmount * BPS / (10000 + negligibleRateDiffBps);
+
+        for (uint i = 0; i < reserveArr.length; i++) {
+
+            if (i == bestReserve.index) continue;
+
+            isPayingFees = isFeePayingReserve[address(reserve)];
+            srcAmountWithFee = ((src == ETH_TOKEN_ADDRESS) && isPayingFees) ? srcAmount - takerFee : srcAmount;
+            destAmount = srcAmountWithFee * rates[i] / PRECISION;
+            destAmount = (dest == ETH_TOKEN_ADDRESS && isPayingFees) ? destAmount * (BPS - takerFee) / BPS : destAmount;
+
+            if (destAmount > bestReserve.destAmount) {
+                reserveCandidates[numRelevantReserves++] = i;
+            }
+        }
+
+        if (numRelevantReserves > 1) {
+            //when encountering small rate diff from bestRate. draw from relevant reserves
+            bestReserve.index = reserveCandidates[uint(blockhash(block.number-1)) % numRelevantReserves];
+        } else {
+            bestReserve.index = reserveCandidates[0];
+        }
+        
+        isPayingFees = isFeePayingReserve[address(reserveArr[bestReserve.index])];
+        return (reserveArr[bestReserve.index], rates[bestReserve.index], isPayingFees);
+    }
+
+    uint constant rateMul = 1;
```

Comment:
What's this lol... Unused constant.

Reply:
ho yes.
So I wanted some multiplier to use for finding the rate in the per reserve array. which holds rate and split BPS
currently I mul by constant.
will probably move to this....

lets discuss monday